test/deployframework/framework.go

```diff
@@ -0,0 +1,335 @@
+	df.Config.Namespace = namespace
+	cfg.Namespace = namespace
+	cfg.MeteringConfig.ObjectMeta = meta.ObjectMeta{
+		Name:      meteringconfigMetadataName,
+		Namespace: namespace,
+	}
+
+	df.Logger.Debugf("Deployer config: %+v", cfg)
+
+	df.Deployer, err = deploy.NewDeployer(cfg, df.Logger, df.Client, df.APIExtClient, df.MeteringClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct a new deployer object: %v", err)
+	}
+
+	err = df.Deployer.Install()
+	if err != nil {
+		return nil, fmt.Errorf("failed to install metering: %v", err)
+	}
+
+	err = df.addTestingNamespaceLabel(namespace, testNamespaceLabel)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add the testing label to the %s namespace", namespace)
+	}
+
+	_, err = df.WaitForMeteringPods(targetPods, cfg.Namespace)
+	if err != nil {
+		return nil, fmt.Errorf("error waiting for metering pods to become ready: %v", err)
+	}
+
+	routeBearerToken, err := df.GetRouteBearerToken(cfg.Namespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the route bearer token: %v", err)
+	}
+
+	reportResultsPath := filepath.Join(testOutputPath, reportResultsDir)
+	err = os.Mkdir(reportResultsPath, 0777)
+	if err != nil {
+		df.Logger.Fatalf("Failed to make the directory %s: %v", reportResultsPath, err)
+	}
+
+	df.Logger.Infof("Report results directory: %s", reportResultsPath)
+
+	// TODO create functions that determine the reportingframework fields
+	// we can hardcode these values for now as we only have one
+	// meteringconfig configuration that gets installed and we dont
+	// support passing a METERING_CR_FILE yaml file for local testing
+	useHTTPSAPI := df.checkForHTTPSAPI(*cfg.MeteringConfig)
+	useRouteForReportingAPI := true
+	useKubeProxyForReportingAPI := false
+	reportingAPIURL := ""
+
+	reportingFrameworkConfig := &ReportingFrameworkConfig{
+		UseKubeProxyForReportingAPI: useKubeProxyForReportingAPI,
+		UseRouteForReportingAPI:     useRouteForReportingAPI,
+		HTTPSAPI:                    useHTTPSAPI,
+		Namespace:                   cfg.Namespace,
+		KubeConfigPath:              df.KubeConfigPath,
+		ReportingAPIURL:             reportingAPIURL,
+		RouteBearerToken:            routeBearerToken,
+		ReportResultsOutputPath:     reportResultsPath,
+	}
+
+	return reportingFrameworkConfig, nil
+}
+
+// Teardown is a method that creates the resource and container logging
+// directories, then populates those directories by executing the
+// df.CleanupScriptPath bash script, while streaming the script output
+// to stdout. Once the cleanup script has finished execution, we can
+// uninstall the metering stack and return an error if there is any.
+func (df *DeployFramework) Teardown(pathToTestCaseOutputDir string) error {
+	logger := df.Logger.WithFields(logrus.Fields{"component": "cleanup"})
+
+	envVarArr, err := df.createResourceDirs(pathToTestCaseOutputDir)
+	if err != nil {
+		logger.Warnf("failed to create the resource output directories: %v", err)
```

Comment:
Probably return the error

Reply:
If I return an error here, the `df.Deployer.Uninstall()` is not run.

Reply:
Okay, does the cleanup script regularly error then or what?

Reply:
So I can either update this method to be purely focused on logging, and move the `df.Deployer.Uninstall()` to the defer closure, e.g.:
```go
	defer func() {
		err := df.Teardown(testOutputDir)
		assert.NoError(t, err, "capturing logs and uninstalling metering should produce no error")

		err := df.Deployer.Uninstall()
		assert.NoError(t, err, ...)
	}()
```

Reply:
@chancez no the cleanup script doesn't error often but was something that I encountered while developing this and placed more value on the resources being torn down properly.

Reply:
I'm just wondering, what if we consider the script to be very important to run too. I would assume we want to know that. Perhaps we should accumulate errors in a slice, then run uninstall, and if there's any errors in the slice, return the list of errors joined together.